File: cs/apply.py
"""`cs apply`: render ~/.claude from the config repo's claude/ directory and
manage the git identity includes.

Linked (symlink):  CLAUDE.md, rules/, agents/, themes/, keybindings.json, skills/<name>/, plans/ (-> repo/plans)
Rendered (file):   settings.json  = settings.base.json (+) settings.<profile>.json... (+) settings.<machine>.json
Rendered (file):   ~/.config/git/claude-share.inc + identity-<id>.inc ; include line in ~/.gitconfig
"""
from __future__ import annotations

import json
import shutil
import time
from pathlib import Path
from typing import Dict, List, Tuple

from . import jsonmerge, paths
from .config import Machine
from .manifest import Manifest
from .ui import info, warn, ok, act

LINK_ITEMS = ["CLAUDE.md", "rules", "agents", "themes", "keybindings.json"]
GIT_INCLUDE_MARK = "# >>> claude-share >>>"
GIT_INCLUDE_END = "# <<< claude-share <<<"


def _backup(target: Path) -> Path:
    bdir = paths.state_dir() / "backups" / time.strftime("%Y%m%d-%H%M%S")
    bdir.mkdir(parents=True, exist_ok=True)
    dest = bdir / target.name
    shutil.move(str(target), str(dest))
    return dest


def _merge_dir_into(src_dir: Path, dst_dir: Path) -> None:
    """Move files from dst_dir into src_dir (existing files in src_dir win)."""
    for f in sorted(dst_dir.rglob("*")):
        if not f.is_file():
            continue
        rel = f.relative_to(dst_dir)
        t = src_dir / rel
        if not t.exists():
            t.parent.mkdir(parents=True, exist_ok=True)
            shutil.move(str(f), str(t))
    shutil.rmtree(dst_dir)


def _link(src: Path, dst: Path, check: bool, changes: List[str]) -> None:
    """Make dst a symlink to src. A pre-existing real file/dir on the machine is
    adopted into the repo (dirs: merged, repo wins on collisions; files: moved
    if the repo has none, else backed up)."""
    if dst.is_symlink():
        if dst.resolve() == src.resolve():
            return
        changes.append(f"relink {paths.contract(dst)}")
        if not check:
            dst.unlink()
            dst.symlink_to(src)
        return
    if dst.exists():
        if dst.is_dir():
            changes.append(f"adopt {paths.contract(dst)} -> {paths.contract(src)} (merge)")
            if not check:
                src.mkdir(parents=True, exist_ok=True)
                _merge_dir_into(src, dst)
                dst.symlink_to(src)
        elif not src.exists():
            changes.append(f"adopt {paths.contract(dst)} -> {paths.contract(src)}")
            if not check:
                src.parent.mkdir(parents=True, exist_ok=True)
                shutil.move(str(dst), str(src))
                dst.symlink_to(src)
        else:
            changes.append(f"replace {paths.contract(dst)} (backup kept)")
            if not check:
                _backup(dst)
                dst.symlink_to(src)
        return
    if not src.exists():
        return
    changes.append(f"link {paths.contract(dst)}")
    if not check:
        dst.parent.mkdir(parents=True, exist_ok=True)
        dst.symlink_to(src)


def settings_layers(repo: Path, m: Machine) -> List[Tuple[str, Dict]]:
    layers: List[Tuple[str, Dict]] = []
    names = ["settings.base.json"] + [f"settings.{p}.json" for p in m.profiles] + [f"settings.{m.name}.json"]
    for n in names:
        f = repo / "claude" / n
        if f.exists():
            layers.append((n, jsonmerge.loads(f.read_text())))
    return layers


def render_settings(repo: Path, m: Machine) -> Dict:
    return jsonmerge.merge_layers(*[d for _, d in settings_layers(repo, m)])


def apply_settings(repo: Path, m: Machine, check: bool, changes: List[str]) -> None:
    target = paths.claude_dir() / "settings.json"
    desired = render_settings(repo, m)
    if not settings_layers(repo, m):
        return
    current = jsonmerge.loads(target.read_text()) if target.exists() else {}
    if current == desired:
        return
    keys = jsonmerge.diff_keys(current, desired)
    changes.append(f"settings.json: {', '.join(keys[:8])}{' ...' if len(keys) > 8 else ''}")
    if not check:
        target.parent.mkdir(parents=True, exist_ok=True)
        if target.exists():
            _backup_copy(target)
        target.write_text(jsonmerge.dumps(desired))


def _backup_copy(target: Path) -> None:
    bdir = paths.state_dir() / "backups" / time.strftime("%Y%m%d-%H%M%S")
    bdir.mkdir(parents=True, exist_ok=True)
    shutil.copy2(str(target), str(bdir / target.name))


def apply_links(repo: Path, check: bool, changes: List[str]) -> None:
    cdir = paths.claude_dir()
    cdir.mkdir(parents=True, exist_ok=True)
    for item in LINK_ITEMS:
        _link(repo / "claude" / item, cdir / item, check, changes)
    # skills: one link per skill dir so marketplace/plugin skills can coexist
    skills_src = repo / "claude" / "skills"
    if skills_src.is_dir():
        (cdir / "skills").mkdir(exist_ok=True)
        for s in sorted(skills_src.iterdir()):
            if s.is_dir():
                _link(s, cdir / "skills" / s.name, check, changes)
    # plans: whole dir
    _link(repo / "plans", cdir / "plans", check, changes)


def render_git_includes(man: Manifest) -> Dict[Path, str]:
    gdir = paths.home() / ".config" / "git"
    files: Dict[Path, str] = {}
    inc_lines = ["# generated by `cs apply` — do not edit; edit projects.toml [identities] instead"]
    for ident in man.identities.values():
        body = [
            f"# identity '{ident.id}' (generated by cs apply)",
            "[user]",
            f"\tname = {ident.name}",
            f"\temail = {ident.email}",
        ]
        if ident.ssh_key:
            key = paths.contract(paths.expand(ident.ssh_key))
            body += ["[core]", f"\tsshCommand = ssh -i {key} -o IdentitiesOnly=yes"]
        files[gdir / f"identity-{ident.id}.inc"] = "\n".join(body) + "\n"
        for g in ident.url_globs:
            inc_lines.append(f'[includeIf "hasconfig:remote.*.url:{g}"]')
            inc_lines.append(f"\tpath = identity-{ident.id}.inc")
    files[gdir / "claude-share.inc"] = "\n".join(inc_lines) + "\n"
    return files


def apply_git(man: Manifest, check: bool, changes: List[str]) -> None:
    for f, content in render_git_includes(man).items():
        if f.exists() and f.read_text() == content:
            continue
        changes.append(f"write {paths.contract(f)}")
        if not check:
            f.parent.mkdir(parents=True, exist_ok=True)
            f.write_text(content)
    gc = paths.home() / ".gitconfig"
    block = f"{GIT_INCLUDE_MARK}\n[include]\n\tpath = ~/.config/git/claude-share.inc\n{GIT_INCLUDE_END}\n"
    text = gc.read_text() if gc.exists() else ""
    if GIT_INCLUDE_MARK in text:
        start, end = text.index(GIT_INCLUDE_MARK), text.index(GIT_INCLUDE_END) + len(GIT_INCLUDE_END) + 1
        new = text[:start] + block + text[end:]
    else:
        new = text + ("\n" if text and not text.endswith("\n") else "") + block
    if new != text:
        changes.append("~/.gitconfig: include claude-share.inc")
        if not check:
            gc.write_text(new)


def run(repo: Path, m: Machine, man: Manifest, check: bool = False) -> List[str]:
    changes: List[str] = []
    apply_settings(repo, m, check, changes)
    apply_links(repo, check, changes)
    apply_git(man, check, changes)
    for c in changes:
        (info if check else act)(c)
    if not changes:
        ok("~/.claude up to date")
    return changes

File: src/apply.ts
/** cs apply: render ~/.claude from <repo>/claude, git identity includes, shell rc block. */
import { existsSync, lstatSync, mkdirSync, readdirSync, readFileSync, readlinkSync, realpathSync, renameSync, rmdirSync, rmSync, statSync, symlinkSync, unlinkSync, writeFileSync, copyFileSync } from "node:fs";
import { basename, dirname, join, resolve } from "node:path";
import { claudeDir, contract, expand, home, stateDir, toolRoot } from "./paths.js";
import { shellRc } from "./platform.js";
import { globs, keyPath, type Manifest } from "./manifest.js";
import type { Share } from "./share.js";
import { diffKeys, dumps, loads, mergeLayers } from "./jsonmerge.js";

const LINK_ITEMS = ["CLAUDE.md", "rules", "agents", "themes", "keybindings.json", "statusline.sh"];
const GIT_MARK = "# >>> claude-share >>>", GIT_END = "# <<< claude-share <<<";
const stamp = () => new Date().toISOString().replace(/[-:]/g, "").slice(0, 15);

function backup(target: string) {
  const d = join(stateDir(), "backups", stamp()); mkdirSync(d, { recursive: true });
  const dest = join(d, basename(target));
  if (statSync(target).isDirectory()) renameSync(target, dest); else { copyFileSync(target, dest); unlinkSync(target); }
}
function mergeDirInto(src: string, dst: string) {
  const walk = (dir: string) => { for (const e of readdirSync(dir, { withFileTypes: true })) { const f = join(dir, e.name); if (e.isDirectory()) walk(f); else {
    const rel = f.slice(dst.length + 1); const t = join(src, rel);
    if (!existsSync(t)) { mkdirSync(dirname(t), { recursive: true }); renameSync(f, t); } } } };
  walk(dst); rmSync(dst, { recursive: true, force: true });
}
const isLink = (p: string) => { try { return lstatSync(p).isSymbolicLink(); } catch { return false; } };
const real = (p: string) => { try { return realpathSync(p); } catch { return resolve(p); } };

function link(src: string, dst: string, check: boolean, changes: string[]) {
  if (isLink(dst)) {
    // a link that resolves to src through another link (e.g. ~/.claude/skills/x → ../../.agents/skills/x → repo) is fine as is
    if (resolve(dirname(dst), readlinkSync(dst)) === resolve(src) || real(dst) === real(src)) return;
    changes.push(`relink ${contract(dst)}`); if (!check) { unlinkSync(dst); symlinkSync(src, dst); } return;
  }
  if (existsSync(dst)) {
    if (statSync(dst).isDirectory()) { changes.push(`import ${contract(dst)} → ${contract(src)} (merge)`); if (!check) { mkdirSync(src, { recursive: true }); mergeDirInto(src, dst); symlinkSync(src, dst); } }
    else if (!existsSync(src)) { changes.push(`import ${contract(dst)} → ${contract(src)}`); if (!check) { mkdirSync(dirname(src), { recursive: true }); renameSync(dst, src); symlinkSync(src, dst); } }
    else { changes.push(`replace ${contract(dst)} (backup kept)`); if (!check) { backup(dst); symlinkSync(src, dst); } }
    return;
  }
  if (!existsSync(src)) return;
  changes.push(`link ${contract(dst)}`); if (!check) { mkdirSync(dirname(dst), { recursive: true }); symlinkSync(src, dst); }
}

export function settingsLayers(share: Share): [string, Record<string, unknown>][] {
  const m = share.machine; const names = ["settings.base.json", ...m.profiles.map((p) => `settings.${p}.json`), `settings.${m.name}.json`];
  return names.filter((n) => existsSync(join(share.path, "claude", n))).map((n) => [n, loads(readFileSync(join(share.path, "claude", n), "utf8"))]);
}
export const renderSettings = (share: Share) => mergeLayers(...settingsLayers(share).map(([, d]) => d));

export function applySettings(share: Share, check: boolean, changes: string[]) {
  if (!settingsLayers(share).length) return;
  const target = join(claudeDir(), "settings.json");
  const desired = renderSettings(share);
  const current = existsSync(target) ? loads(readFileSync(target, "utf8")) : {};
  if (JSON.stringify(current) === JSON.stringify(desired)) return;
  const keys = diffKeys(current, desired);
  changes.push(`settings.json: ${keys.slice(0, 8).join(", ")}${keys.length > 8 ? " …" : ""}`);
  if (!check) { mkdirSync(claudeDir(), { recursive: true }); if (existsSync(target)) { const d = join(stateDir(), "backups", stamp()); mkdirSync(d, { recursive: true }); copyFileSync(target, join(d, "settings.json")); } writeFileSync(target, dumps(desired)); }
}
export function applyLinks(repo: string, check: boolean, changes: string[]) {
  const cdir = claudeDir(); mkdirSync(cdir, { recursive: true });
  for (const item of LINK_ITEMS) link(join(repo, "claude", item), join(cdir, item), check, changes);
  // skills.sh (`npx skills add … -g`) installs into ~/.agents/skills and symlinks ~/.claude/skills/<name> there;
  // make that directory (and its lock file) the repo's claude/skills so installs land in the share directly.
  const skills = join(repo, "claude", "skills"), agents = join(home(), ".agents");
  link(skills, join(agents, "skills"), check, changes);
  link(join(repo, "claude", "skill-lock.json"), join(agents, ".skill-lock.json"), check, changes);
  if (existsSync(skills)) { mkdirSync(join(cdir, "skills"), { recursive: true }); for (const e of readdirSync(skills, { withFileTypes: true })) if (e.isDirectory()) link(join(skills, e.name), join(cdir, "skills", e.name), check, changes); }
  link(join(repo, "plans"), join(cdir, "plans"), check, changes);
}
export function renderGitIncludes(man: Manifest): Record<string, string> {
  const gdir = join(home(), ".config", "git"); const files: Record<string, string> = {};
  const inc = ["# generated by `cs apply` — do not edit; edit projects.toml [identities] instead"];
  for (const i of Object.values(man.identities)) {
    files[join(gdir, `identity-${i.id}.inc`)] = [`# identity '${i.id}' (generated by cs apply)`, "[user]", `\tname = ${i.name}`, `\temail = ${i.email}`, "[core]",
      `\tsshCommand = ssh -i ${contract(expand(keyPath(i)))} -o IdentitiesOnly=yes`].join("\n") + "\n";
    for (const g of globs(i)) inc.push(`[includeIf "hasconfig:remote.*.url:${g}"]`, `\tpath = identity-${i.id}.inc`);
  }
  files[join(gdir, "claude-share.inc")] = inc.join("\n") + "\n";
  return files;
}
export function applyGit(man: Manifest, check: boolean, changes: string[]) {
  const gdir = join(home(), ".config", "git"); const wanted = renderGitIncludes(man);
  if (existsSync(gdir)) for (const f of readdirSync(gdir)) if (/^identity-.*\.inc$/.test(f) && !(join(gdir, f) in wanted)) { changes.push(`remove stale ${contract(join(gdir, f))}`); if (!check) unlinkSync(join(gdir, f)); }
  for (const [f, content] of Object.entries(wanted)) { if (existsSync(f) && readFileSync(f, "utf8") === content) continue; changes.push(`write ${contract(f)}`); if (!check) { mkdirSync(gdir, { recursive: true }); writeFileSync(f, content); } }
  const gc = join(home(), ".gitconfig"); const text = existsSync(gc) ? readFileSync(gc, "utf8") : "";
  const block = `${GIT_MARK}\n[include]\n\tpath = ~/.config/git/claude-share.inc\n${GIT_END}\n`;
  const next = text.includes(GIT_MARK) ? text.slice(0, text.indexOf(GIT_MARK)) + block + text.slice(text.indexOf(GIT_END) + GIT_END.length + 1) : text + (text && !text.endsWith("\n") ? "\n" : "") + block;
  if (next !== text) { changes.push("~/.gitconfig: include claude-share.inc"); if (!check) writeFileSync(gc, next); }
}
export function applyShellRc(check: boolean, changes: string[]) {
  const rc = shellRc(); const sh = contract(join(toolRoot(), "shell", "cs.sh"));
  const block = `${GIT_MARK}\n[ -f "${sh}" ] && . "${sh}"\n${GIT_END}\n`;
  const text = existsSync(rc) ? readFileSync(rc, "utf8") : "";
  const next = text.includes(GIT_MARK) ? text.slice(0, text.indexOf(GIT_MARK)) + block + text.slice(text.indexOf(GIT_END) + GIT_END.length + 1) : text + (text && !text.endsWith("\n") ? "\n" : "") + block;
  if (next !== text) { changes.push(`${contract(rc)}: source shell/cs.sh (claude() wrapper, PATH)`); if (!check) writeFileSync(rc, next); }
}
/** Everything ~/.claude, the git includes and the shell rc need to match the share (silent: the change lines come back,
 *  the caller prints them); with `check` nothing is written. */
export function runApply(share: Share, check = false): string[] {
  const changes: string[] = [];
  applySettings(share, check, changes); applyLinks(share.path, check, changes); applyGit(share.manifest, check, changes); applyShellRc(check, changes);
  return changes;
}
